entity_gym/environment.py

```diff
@@ -123,6 +124,7 @@ def state_space(cls) -> Dict[str, Entity]:
 
     @classmethod
     @abstractmethod
+    
```

Comment:
Why do we have a baked in dictionary into our action space?  Shouldn't we instead just have our action space be directly an ActionSpace object, where the ActionSpace itself may be a UnionActionSpace[FirstActionSpace, SecondActionSpace,...]?  The additional str identifiers everywhere make the API harder to parse.  At the least, I think switching to a newtype style type (as you did for EntityID) for all the string identifiers would make the API easier to read.

Reply:
Good point, also not a fan of stringly typed code. There needs to be some way to tie identify and tie together action spaces/action heads/action masks (e.g. [this code](https://github.com/cswinter/entity-gym/blob/2fa93df66996907b1771291a0d11e8243ab4d857/entity_gym/main.py#L68)). Initially, I had a base class for actions with a `name` but that turned out to be kind of annoying and I just decided to move it into a dict so it's consistent with the state space as well. 
Some alternatives, curious to get your thoughts on those or any ideas you might have:
1. return a `List[ActionSpace]`, instead of a string identifier the primary key is the index
2. return an `ActionSpace`, and add composite `ActionSpace`s that allow you to compose base `ActionSpace`s
I think we do want some way of associating human-readable names associated with actions for debugging/metrics/manually playable envs. For (1), we could also (optionally) return a corresponding `List[str]` of labels, and composite `ActionSpace`s could support something similar.

Reply:
There's a somewhat related annoyance that for a given environment the actions/action masks/action space have a specific and related structure that isn't captured by the generic interface. So you end up with a bunch of string dict lookups and type asserts, e.g. https://github.com/cswinter/entity-gym/blob/2fa93df66996907b1771291a0d11e8243ab4d857/entity_gym/envs/pick_matching_balls.py#L85-L86.

We could solve this with associate types on `Environment`/`ActionSpace` (I think this is not actually expressible in MyPy):

```rust
trait ActionSpace {
    type TActionMask: ActionMask;
    type TAction: Action;
}

struct CategoricalActionSpace {
    actions: Vec<String>,
}

impl ActionSpace for CategoricalActionSpace {
    type TActionMask = CategoricalActionMask;
    type TAction = CategoricalAction;
}

trait Environment {
    type TActionSpace: ActionSpace

    fn action_space() -> ActionSpace;

    fn act(&mut self, actions: Self::ActionSpace::TAction) -> (Observation, Self::ActionSpace::ActionMask);
}
```

Seems like a bad idea though unless we want this library to be used only by Haskell fanatics 😛 Composite action spaces would also need something like HList/HDict. Maybe there's a simpler or more Pythonic solution that captures some of the benefits.